PyCBC threshold code is broken on intel mac with openMP

See discussion here https://github.com/conda-forge/pycbc-feedstock/pull/84. There is some issue with the openmp code used for the threshold code, which appears to manifest on old intel macs. This is covered (thankfully) by our unit tests, but it is possible that whatever is causing this will appear in other contexts in the future. We should understand and fix ... But we need to be able to reproduce it first!

Sorry, apparently I missed that @josh-willis had a suggestion.

Why do we need two separate `#pragma` directives? The compiler does not seem to like it:
```
      In file included from pycbc/events/simd_threshold_cython.cpp:1286:
      pycbc/events/simd_threshold_ccode.cpp:95:13: error: exactly one 'ordered' directive must appear in the loop body of an enclosing directive
                  #pragma omp ordered
                  ^
      pycbc/events/simd_threshold_ccode.cpp:91:13: note: previous 'ordered' directive used here
                  #pragma omp ordered
```
However if I do this:
```diff
diff --git a/pycbc/events/simd_threshold_ccode.cpp b/pycbc/events/simd_threshold_ccode.cpp
index ae5910e5e..51ea2a931 100644
--- a/pycbc/events/simd_threshold_ccode.cpp
+++ b/pycbc/events/simd_threshold_ccode.cpp
@@ -91,10 +91,9 @@ void _parallel_threshold(int64_t N, std::complex<float> * __restrict arr,
             #pragma omp ordered
             {
                 t+=c;
+                memmove(outl+t-c, outl+start, sizeof(unsigned int)*c);
+                memmove(outv+t-c, outv+start, sizeof(std::complex<float>)*c);
             }
-            memmove(outl+t-c, outl+start, sizeof(unsigned int)*c);
-            memmove(outv+t-c, outv+start, sizeof(std::complex<float>)*c);
-
         }
 
         count[0] = t;
```
Then indeed the test passes:
```
$ python test/test_threshold.py                    
========================================================================
Running CPU unit tests for Threshold:
test_threshold (__main__.TestThreshold.test_threshold) ... Reference: 14327 locs, 14327 vals
Test: 14327 locs (0 different), 14327 vals (0 different)
locs in [0, 65536) are all equal (916 values)
locs in [65536, 131072) are all equal (862 values)
locs in [131072, 196608) are all equal (916 values)
locs in [196608, 262144) are all equal (858 values)
locs in [262144, 327680) are all equal (926 values)
locs in [327680, 393216) are all equal (917 values)
locs in [393216, 458752) are all equal (874 values)
locs in [458752, 524288) are all equal (864 values)
locs in [524288, 589824) are all equal (924 values)
locs in [589824, 655360) are all equal (919 values)
locs in [655360, 720896) are all equal (874 values)
locs in [720896, 786432) are all equal (875 values)
locs in [786432, 851968) are all equal (887 values)
locs in [851968, 917504) are all equal (874 values)
locs in [917504, 983040) are all equal (924 values)
locs in [983040, 1048576) are all equal (917 values)
ok

----------------------------------------------------------------------
Ran 1 test in 0.133s

OK
```

--- a/pycbc/events/simd_threshold_ccode.cpp
+++ b/pycbc/events/simd_threshold_ccode.cpp
@@ -92,11 +92,14 @@ void _parallel_threshold(int64_t N, std::complex<float> * __restrict arr,
 
             #pragma omp ordered
             {
-                t+=c;
+                t += c;
+                memmove(
+                    outl + t - c, outl + start, sizeof(unsigned int)*c
+                );
+                memmove(
+                    outv + t - c, outv + start, sizeof(std::complex<float>)*c
+                );
             }
-            memmove(outl+t-c, outl+start, sizeof(unsigned int)*c);
-            memmove(outv+t-c, outv+start, sizeof(std::complex<float>)*c);
-
         }
 
         count[0] = t;

--- a/test/test_threshold.py
+++ b/test/test_threshold.py
@@ -14,47 +14,41 @@
 # with this program; if not, write to the Free Software Foundation, Inc.,
 # 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 
-#
-# =============================================================================
-#
-#                                   Preamble
-#
-# =============================================================================
-#
 """
-These are the unittests for the pycbc.waveform module
+Unit tests for PyCBC's thresholding code.
 """
+
 import unittest
 import numpy
-from pycbc.types import *
-from pycbc.scheme import *
-from pycbc.events import *
+from pycbc.types import Array, complex64
+from pycbc.events import threshold
 from utils import parse_args_all_schemes, simple_exit
 
 _scheme, _context = parse_args_all_schemes("Threshold")
 
-from pycbc.events.threshold_cpu import threshold_numpy
-trusted_threshold = threshold_numpy
+from pycbc.events.threshold_cpu import threshold_numpy as trusted_threshold
+
 
 class TestThreshold(unittest.TestCase):
-    def setUp(self,*args):
+    def setUp(self, *args):
         self.context = _context
         self.scheme = _scheme
         r = numpy.random.uniform(low=-1, high=1.0, size=2**20)
         i = numpy.random.uniform(low=-1, high=1.0, size=2**20)
-        v = r + i*1.0j
+        v = r + i * 1.0j
         self.series = Array(v, dtype=complex64)
         self.threshold = 1.3
         self.locs, self.vals = trusted_threshold(self.series, self.threshold)
-        self.tolerance = 1e-6
-        print(len(self.locs), len(self.vals))
+        print(f'Reference: {len(self.locs)} locs, {len(self.vals)} vals')
 
     def test_threshold(self):
         with self.context:
             locs, vals = threshold(self.series, self.threshold)
+            print(f'Test: {len(locs)} locs, {len(vals)} vals')
             self.assertTrue((locs == self.locs).all())
             self.assertTrue((vals == self.vals).all())
-            print(len(locs), len(vals))
+
+
 suite = unittest.TestSuite()
 suite.addTest(unittest.TestLoader().loadTestsFromTestCase(TestThreshold))
 

